plugin/pkg/admission/ingressclass/admission.go

```diff
@@ -0,0 +1,150 @@
+
+// newPlugin creates a new admission plugin.
+func newPlugin() *claimDefaulterPlugin {
+	return &claimDefaulterPlugin{
+		Handler: admission.NewHandler(admission.Create),
+	}
+}
+
+// SetExternalKubeInformerFactory sets a lister and readyFunc for this
+// claimDefaulterPlugin using the provided SharedInformerFactory.
+func (a *claimDefaulterPlugin) SetExternalKubeInformerFactory(f informers.SharedInformerFactory) {
+	informer := f.Networking().V1beta1().IngressClasses()
+	a.lister = informer.Lister()
+	a.SetReadyFunc(informer.Informer().HasSynced)
+}
+
+// ValidateInitialization ensures lister is set.
+func (a *claimDefaulterPlugin) ValidateInitialization() error {
+	if a.lister == nil {
+		return fmt.Errorf("missing lister")
+	}
+	return nil
+}
+
+// Admit sets the default value of a Ingress's class if the user did not specify
+// a class.
+func (a *claimDefaulterPlugin) Admit(ctx context.Context, attr admission.Attributes, o admission.ObjectInterfaces) error {
+	if attr.GetResource().GroupResource() != networkingv1beta1.Resource("ingresses") {
+		return nil
+	}
+
+	if len(attr.GetSubresource()) != 0 {
+		return nil
+	}
+
+	ingress, ok := attr.GetObject().(*networking.Ingress)
+	// if we can't convert then we don't handle this object so just return
+	if !ok {
+		klog.V(3).Infof("Expected Ingress resource, got: %v", attr.GetKind())
+		return nil
+	}
+
+	// IngressClassName field has been set, no need to set a default value.
+	if ingress.Spec.IngressClassName != nil {
+		return nil
+	}
+
+	// Ingress class annotation has been set, no need to set a default value.
+	if _, ok := ingress.Annotations[networkingv1beta1.AnnotationIngressClass]; ok {
+		return nil
+	}
+
+	klog.V(4).Infof("No class specified on Ingress %s", ingress.Name)
+
+	defaultClass, err := getDefaultClass(a.lister)
+	if err != nil {
+		return admission.NewForbidden(attr, err)
+	}
+
+	// No default class specified, no need to set a default value.
+	if defaultClass == nil {
+		return nil
+	}
+
+	klog.V(4).Infof("Defaulting class for Ingress %s to %s", ingress.Name, defaultClass.Name)
+	ingress.Spec.IngressClassName = &defaultClass.Name
+	return nil
+}
+
+// getDefaultClass returns the default IngressClass from the store, or nil.
+func getDefaultClass(lister networkingv1beta1listers.IngressClassLister) (*networkingv1beta1.IngressClass, error) {
+	list, err := lister.List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+
+	defaultClasses := []*networkingv1beta1.IngressClass{}
+	for _, class := range list {
+		if class.Annotations[networkingv1beta1.AnnotationIsDefaultIngressClass] == "true" {
+			defaultClasses = append(defaultClasses, class)
+		}
+	}
+
+	if len(defaultClasses) == 0 {
+		return nil, nil
+	}
+
+	if len(defaultClasses) > 1 {
+		klog.V(3).Infof("More than %d IngressClasses marked as default (only 1 allowed)", len(defaultClasses))
+		return nil, errors.NewInternalError(fmt.Errorf("%d default IngressClasses were found, only 1 allowed", len(defaultClasses)))
```

Comment:
Agree that neither of those are ideal, but I can't think of a better option. I think the 2nd option is reasonable, all things considered.

Reply:
if multiple indicate they are a default, we could take the first lexicographically (to be deterministic)

Reply:
I think one potential downside of that approach would be that it would hide an underlying configuration error. I don't think two default ingress classes should ever be something considered valid. I can imagine that approach resulting in very confusing behavior for an end user in the following scenario:

1. An `abc` ingress class is marked as default
2. Sometime later, a new `def` ingress class is marked as default
3. That default I just set isn't working - did I set the annotation right? Maybe I should file a GitHub issue...

I think it's pretty straightforward to add and remove an annotation, so I don't think there will be a huge amount of time involved in changing a default ingress class. In those cases where one step is missed, I think it's better to return an error message than to hide an error message by providing what would likely be unexpected behavior.

Reply:
For what it's worth, this is closely modeled after the StorageClass admission controller, even the name of the annotation matches. That admission controller also returns an error if there are more than 1 classes set as default.

Reply:
ok